hobbes: init at latest

thmzlt opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change

Introducing the [hobbes](morgan-stanley/hobbes) programming and execution environment build and maintained at Morgan Stanley.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

thmzlt commented:
Looking into the build timeout. [It passes](https://travis-ci.org/Morgan-Stanley/hobbes/jobs/642605695) on Travis CI.

doronbehar commented on pkgs/development/tools/hobbes/default.nix:
```diff
@@ -0,0 +1,38 @@
+      Hobbes is a a language, embedded compiler, and runtime for efficient
+      dynamic expression evaluation, data storage and analysis.
+    '';
+    homepage = https://github.com/Morgan-Stanley/hobbes;
```
```suggestion
    homepage = "https://github.com/Morgan-Stanley/hobbes";
```

doronbehar commented on pkgs/development/tools/hobbes/default.nix:
```diff
@@ -0,0 +1,38 @@
+{ stdenv, pkgs }:
```
We usually import all the necessary arguments not with `pkgs` but with each package showing up in the list here.

doronbehar commented on pkgs/development/tools/hobbes/default.nix:
```diff
@@ -0,0 +1,38 @@
+    zlib
+  ];
+
+  doCheck = true;
```
Didn't the builder ran `make check` or whatever without this?

thmzlt replied:
Good point. It probably didn't run because the checkTarget was incorrect.

thmzlt replied:
Builder didn't run the tests without this flag (even with the right target).

doronbehar commented on pkgs/development/tools/hobbes/default.nix:
```diff
@@ -0,0 +1,38 @@
+  ];
+
+  doCheck = true;
+  checkTarget = "test";
```
Usually this is used by default.

thmzlt replied:
If I removed it, the tests don't run. Isn't the default target "check"?

doronbehar replied:
Correct, my bad.

doronbehar commented on pkgs/development/tools/hobbes/default.nix:
```diff
@@ -0,0 +1,38 @@
+
+stdenv.mkDerivation {
+  name = "hobbes";
+  version = "latest-432d6a2";
```
```suggestion
  version = "unstable-2020-02-18";
```

thmzlt replied:
Done.

aanderse replied:
I don't see this as being done yet.

thmzlt replied:
My bad, fixing it right away.

doronbehar reviewed (commented):
Besides that all LGTM.

thmzlt commented:
What else can I do to help with this PR? Do I need to tick all the boxes on the template list?

doronbehar commented:
You can try and write at https://discourse.nixos.org/t/prs-already-reviewed/2617 . I don't have merge permissions so I've done as best as I could. Since I can see this is your first contribution, 1st of all welcome aboard :) And 2ndly, get used to it - PRs can take a longgg of time.

thmzlt commented:
Thanks @doronbehar!

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/106

danieldk commented on pkgs/development/tools/hobbes/default.nix:
```diff
@@ -0,0 +1,38 @@
+    sha256 = "0av6wp3gsjnil2j8xfn3rkx2r50sfcaq19d7a8bpnfhq2z8q67kg";
+  };
+
+  nativeBuildInputs = [
```
I don't know this project, but `ncurses`, `readline`, and `zlib` should most likely be in `buildInputs`. From a quick glance at their `CMakeList.txt`, `llvm_6` should probably also be in`buildInputs`. If Python is only used during testing (as it appears), it should be in `checkInputs`. Also, did you try if this works with Python 3? Python 2 is EOL.

thmzlt replied:
Pushed changes reflecting your suggestions (and squashed the other commits). It does work with Python 3. Does it look ok?

danieldk replied:
Did you try to put Python in `checkInputs`? Looks ok otherwise. Can't do a proper review though, since I can't test this package.

thmzlt replied:
Just did that. Indeed Python is only used to test the implementation of the fregion script.

aanderse commented:
@thmzlt after reading the upstream guide to building it seems like more recent `gcc` and `llvm` should work... did you try?
I believe @danieldk has a comment about `python` and `checkInputs` which hasn't been addressed yet.

thmzlt commented:
> @thmzlt after reading the upstream guide to building it seems like more recent gcc and llvm should work... did you try?

Building with latest GCC and LLVM fails. There are open issues on upstream for that.

> I believe @danieldk has a comment about python and checkInputs which hasn't been addressed yet.

I'll fix it later today when I have a chance and let you know. Thank you all for guiding me on this PR.

aanderse commented:
@thmzlt awesome! Links to upstream issues always appreciated so future updates can easily check if these issues are resolved or not.

thmzlt commented:
Pushed changes to set Python as a check phase input. Should be good to go.

Regarding the issues related to LLVM/GCC versions, they are:

- https://github.com/Morgan-Stanley/hobbes/issues/315: LLVM is used as a JIT compiler and there are breaking changes in its API between releases
- https://github.com/Morgan-Stanley/hobbes/issues/316: Warnings (that fail the build) are triggered by GCC 8

Both issues are expected to be fixed.

danieldk commented:
> Pushed changes to set Python as a check phase input. Should be good to go.

Cool, thanks!
 
> Regarding the issues related to LLVM/GCC versions, they are:

Maybe it's worthwhile adding a short comment to the derivation above the respective dependencies with the links? Avoid that somebody has to hunt for PRs in the future to find these links ;).

aanderse commented on pkgs/development/tools/hobbes/default.nix:
```diff
@@ -0,0 +1,44 @@
+
+  nativeBuildInputs = [
+    cmake
+    gcc6
```
Should be removed, see above.

aanderse reviewed (changes requested):
I notice there are a few unresolved comments from other reviewers.

thmzlt commented:
@aanderse, just pushed a new version of the derivation with your feedback. Let me know if there's anything else.

aanderse commented on pkgs/development/tools/hobbes/default.nix:
```diff
@@ -0,0 +1,43 @@
+    python37
+  ];
+
+  doCheck = false;
```
I thought `python` was used for tests but now this is disabled.

aanderse commented on maintainers/maintainer-list.nix:
```diff
@@ -3976,6 +3976,12 @@
     githubId = 10544;
     name = "Giuluo Eulisse";
   };
+  kthielen = {
```
Please separate this into a separate commit. So I'm asking for you to add yourself to the maintainer list in the forest commit, followed by the rest of this PR in your second commit.

aanderse reviewed (changes requested):
This is looking really good! Very close, a few more changes, a quick test, and I'm more than happy to merge. Thanks for your patience up to this point.

Commit:
maintainers: add thmzlt

Commit:
hobbes: init at unstable-2020-03-10

thmzlt commented:
@aanderse, I have fixed the version name, and separated the changes in two commits. I have also changed the maintiner from the upstream's maintainer to myself since I'm the one working on the derivation. Finally, I have added a comment pointing to the upstream issue on why doCheck is currently set to false, and removed the python37 input.

aanderse reviewed (approved):
@thmzlt what a fantastic job you've done here. Welcome to the maintainers team, we're so glad to have you! :tada:

aanderse commented:
@GrahamcOfBorg build hobbes

aanderse commented:
Oh no! :scream: `x86_64-darwin` failure!

thmzlt commented:
I can't test it on Darwin at the moment (I don't have a supported computer). Is it possible to mark this as Linux-only while I figure out how to build on a Mac?

doronbehar commented:
Yes. If upstream intended this to build on darwin, then we mark this as broken there. I.e use `meta.broken = stdenv.isDarwin;`.

If upstream intended this to work only on Linux, you can use the `platforms` meta attribute.

Commit:
hobbes: mark as broken on Darwin

thmzlt commented:
@GrahamcOfBorg build hobbes

thmzlt commented:
Looks like we are good. Thank you all for the help.
